File: src/data_rover/core/table/cells.py
"""Per-cell evaluation. Core produces element IDS (not `TreeItem` — that is an
API-layer projection over the API's own tree-item shape); the route maps ids to
`TreeItem`. Cell kind is derived from column kind x mode x source arity:

| column     | mode     | source arity | cell                                    |
|------------|----------|---------------|-----------------------------------------|
| element    | n/a      | single        | `ElementCell`                           |
| property   | collapse | single        | `ValueCell` (editable iff declared)     |
| property   | collapse | many          | `ValuesCell` (read-only, joined)        |
| property   | expand   | single (req.) | `ValueCell` (read-only — see below)     |
| navigation | collapse | any           | `ElementsCell` (capped for display)     |
| navigation | expand   | single (req.) | `ElementCell` (one binding per row)     |

An `expand` column already promoted its value(s) into a `RowKey` slot back in
`build_rows` (Task 4) — one row per value. Re-deriving that slot's value here
(rather than re-navigating/re-reading the property) is what makes an expand
cell agree with the row it belongs to, including for `keep_empty` rows whose
slot is `None`. `_expand_slot_of` (the SAME helper `build_rows` and
`resolve_source_elements` use — there is exactly one slot-arithmetic
implementation, in `evaluate.py`) finds that slot from the column's own index.

`present`/`editable` for a (collapse, single-element) property cell:
- `present=False`: the source element's TYPE does not declare the property —
  greyed in the UI, never editable (there is nothing to edit).
- `present=True, value=None`: the type declares it but this element has not
  set it — editable (the user can fill it in).
- `editable` is `True` only when the source is single-binding, collapse mode,
  AND the property is present; an `expand` cell is never editable (its slot's
  value came from a promoted binding, not this row's single element in the
  sense the edit UI needs — editing it would mean editing a different row's
  worth of expansion), and a many-element collapse cell is a read-only joined
  view of everyone's values.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from data_rover.core.script.embed import ScriptEvalContext

from data_rover.core.metamodel.schema import Metamodel
from data_rover.core.model.model import Model
from data_rover.core.model.naming import display_name
from data_rover.core.navigation.evaluate import PropertyValue

from .evaluate import (
    Binding,
    RowKey,
    TableLimits,
    _expand_slot_of,
    _navigation_reached,
    resolve_source_elements,
)
from .schema import (
    ElementColumn,
    NavigationColumn,
    PropertyColumn,
    ScriptColumn,
    TableDefinition,
)


@dataclass
class ElementCell:
    element_id: str | None


@dataclass
class ValueCell:
    present: bool
    value: object
    element_id: str | None
    editable: bool


@dataclass
class ValuesCell:
    present: bool
    values: list[object]
    total: int
    truncated: bool


@dataclass
class ElementsCell:
    element_ids: list[str]
    total: int
    truncated: bool


@dataclass
class ErrorCell:
    """A script cell that failed (snippet raised, timed out, budget spent,
    runner unavailable, dangling ref). `message` is short; `traceback` (guest
    frames only) rides along for hover detail. Sorts with empties (last)."""

    message: str
    traceback: str | None = None


@dataclass
class PendingCell:
    """A script cell whose value is not computed yet (cache-only miss, spec
    §4.2): a background sweep is filling it. Renders as a placeholder, never
    as an error; sorts with empties; exports as #ERROR only in the
    failed-sweep path (a completed sweep leaves no pending cells)."""


Cell = ElementCell | ValueCell | ValuesCell | ElementsCell | ErrorCell | PendingCell


def _prop_present(mm: Metamodel, type_name: str, prop: str) -> bool:
    return any(pd.name == prop for pd in mm.effective_element_properties(type_name))


def expand_property_values(
    model: Model, col: PropertyColumn, roots: list[str]
) -> list[Binding]:
    """Values an `expand` property column contributes for ONE row's roots.

    The schema guarantees an expand property column has a single-binding
    source, so `roots` holds at most one element id here.

    A single-valued (or undeclared) property is NOT an error: its scalar value
    expands to exactly one row, so mixed scopes (some types multi-valued, some
    single) and stale configs keep evaluating instead of 422-ing the whole
    table. The UI additionally greys the split toggle out when it can prove no
    scoped type is multi-valued — this is the tolerant evaluation half.
    """
    if not roots:
        return []
    eid = roots[0]
    el = model.elements[eid]
    raw = el.properties.get(col.name)
    if raw is None:
        return []
    return list(raw) if isinstance(raw, (list, tuple)) else [raw]


def _element_cell(
    mm: Metamodel,
    model: Model,
    defn: TableDefinition,
    key: RowKey,
    col: ElementColumn,
    base_slots: int,
    limits: TableLimits,
    script: ScriptEvalContext | None = None,
) -> ElementCell:
    els = resolve_source_elements(
        mm, model, defn, key, col.source, base_slots, limits, script=script
    )
    return ElementCell(element_id=els[0] if els else None)


def _property_cell(
    mm: Metamodel,
    model: Model,
    defn: TableDefinition,
    key: RowKey,
    col: PropertyColumn,
    col_index: int,
    base_slots: int,
    limits: TableLimits,
    script: ScriptEvalContext | None = None,
) -> Cell:
    els = resolve_source_elements(
        mm, model, defn, key, col.source, base_slots, limits, script=script
    )
    if col.mode == "expand":
        # The value already sits in this row's key slot (build_rows promoted
        # it); read it back rather than re-deriving it, so a `keep_empty` row
        # (slot == None) and a genuine value agree with the row they're in.
        slot = _expand_slot_of(defn, base_slots, col_index)
        val = key[slot]
        eid = els[0] if els else None
        present = eid is not None and _prop_present(
            mm, model.elements[eid].type_name, col.name
        )
        return ValueCell(present=present, value=val, element_id=eid, editable=False)
    if not els:
        return ValueCell(present=False, value=None, element_id=None, editable=False)
    if len(els) == 1:
        eid = els[0]
        el = model.elements[eid]
        present = _prop_present(mm, el.type_name, col.name)
        val = el.properties.get(col.name) if present else None
        return ValueCell(present=present, value=val, element_id=eid, editable=present)
    # many-element collapse → joined read-only values (undeclared-on-some-types
    # elements simply contribute nothing, rather than failing the whole cell)
    vals: list[object] = []
    for eid in els:
        el = model.elements[eid]
        if not _prop_present(mm, el.type_name, col.name):
            continue
        v = el.properties.get(col.name)
        if isinstance(v, (list, tuple)):
            vals.extend(v)
        elif v is not None:
            vals.append(v)
    return ValuesCell(present=True, values=vals, total=len(vals), truncated=False)


def _navigation_cell(
    mm: Metamodel,
    model: Model,
    defn: TableDefinition,
    key: RowKey,
    col: NavigationColumn,
    col_index: int,
    base_slots: int,
    limits: TableLimits,
    script: ScriptEvalContext | None = None,
) -> Cell:
    if col.mode == "expand":
        slot = _expand_slot_of(defn, base_slots, col_index)
        b = key[slot]
        if isinstance(b, PropertyValue):
            # The promoted binding is a scalar-property-step terminal: render
            # the VALUE (read-only — it belongs to the element the navigation
            # stepped through, not to this row's single source element).
            return ValueCell(
                present=True, value=b.value, element_id=None, editable=False
            )
        return ElementCell(element_id=b if isinstance(b, str) else None)
    roots = resolve_source_elements(
        mm, model, defn, key, col.source, base_slots, limits, script=script
    )
    reached = _navigation_reached(mm, model, col, roots, limits)
    # `cell_cap` is a per-column display preference; `max_cell_elements` is the
    # server-wide ceiling, so the effective cap is whichever is stricter. An
    # export sets `ignore_cell_caps` so the workbook carries the COMPLETE
    # reached set — a display preference must not drop exported data.
    cap = (
        limits.max_cell_elements
        if limits.ignore_cell_caps
        else min(col.cell_cap, limits.max_cell_elements)
    )
    if any(isinstance(n, PropertyValue) for n in reached):
        # The projected step is a scalar property step: the cell shows VALUES.
        # A mixed frontier (the same property name element-typed on one type,
        # scalar on another) degrades element nodes to their display names so
        # nothing silently drops.
        vals: list[object] = [
            n.value if isinstance(n, PropertyValue) else display_name(model.elements[n])
            for n in reached
        ]
        return ValuesCell(
            present=True,
            values=vals[:cap],
            total=len(vals),
            truncated=len(vals) > cap,
        )
    ids = [n for n in reached if isinstance(n, str)]
    return ElementsCell(element_ids=ids[:cap], total=len(ids), truncated=len(ids) > cap)


def _script_cell(
    mm: Metamodel,
    model: Model,
    defn: TableDefinition,
    key: RowKey,
    col: ScriptColumn,
    col_index: int,
    base_slots: int,
    limits: TableLimits,
    script: ScriptEvalContext | None,
) -> Cell:
    """Cell for one row of a `ScriptColumn`. `expand` reads back the binding
    `build_rows` already promoted into this row's key slot (never re-calls
    `value()`, mirroring `_navigation_cell`/`_property_cell`'s expand
    branches); `collapse` calls fresh (memoized by `script`, so this agrees
    bit-for-bit with any earlier collapse-filter/chaining call made against
    the same `(code, entry, roots)` during `build_rows_ex`)."""
    if col.mode == "expand":
        # The binding already sits in this row's key slot (build_rows promoted
        # it): PropertyValue = a returned scalar; str = a returned element id;
        # None = keep_empty row OR the single error row _expand_values left —
        # re-derive from the memoized call to tell the two apart.
        slot = _expand_slot_of(defn, base_slots, col_index)
        b = key[slot]
        if isinstance(b, PropertyValue):
            return ValueCell(
                present=True, value=b.value, element_id=None, editable=False
            )
        if isinstance(b, str):
            return ElementCell(element_id=b)
        if col.snippet.ref is not None:
            return ErrorCell(message=f"snippet artifact {col.snippet.ref!r} not found")
        if col.snippet.definition is not None and script is not None:
            roots = resolve_source_elements(
                mm, model, defn, key, col.source, base_slots, limits, script=script
            )
            if roots:
                # Force cache-only: the row shape is unknown here (this is a
                # re-derive, not the original build-time call), so a live call
                # could not change the single-row rendering anyway; recomputing
                # it live would also bypass sweep accounting.
                res = script.call(
                    col.snippet.definition.code, "value", roots, cache_only=True
                )
                if res.error is not None:
                    if res.error.kind == "pending":
                        return PendingCell()
                    return ErrorCell(
                        message=res.error.message, traceback=res.error.traceback
                    )
        return ValueCell(present=False, value=None, element_id=None, editable=False)

    if col.snippet.ref is not None:  # post-resolve: dangling ref
        return ErrorCell(message=f"snippet artifact {col.snippet.ref!r} not found")
    if col.snippet.definition is None:  # unconfigured ({}): empty, like an
        # unconfigured navigation/property source
        return ValueCell(present=False, value=None, element_id=None, editable=False)
    els = resolve_source_elements(
        mm, model, defn, key, col.source, base_slots, limits, script=script
    )
    if not els:
        return ValueCell(present=False, value=None, element_id=None, editable=False)
    if script is None:
        # Defensive: routes always supply a context when table_has_script().
        return ErrorCell(message="script runner unavailable")
    res = script.call(col.snippet.definition.code, "value", els)
    if res.error is not None:
        if res.error.kind == "pending":
            return PendingCell()
        return ErrorCell(message=res.error.message, traceback=res.error.traceback)
    p = res.value
    assert p is not None  # CallResult invariant: value xor error
    if p["kind"] == "scalar":
        if p["value"] is None:
            return ValueCell(present=False, value=None, element_id=None, editable=False)
        return ValueCell(
            present=True, value=p["value"], element_id=None, editable=False
        )
    if p["kind"] == "scalars":
        vals = p["values"]
        cap = limits.max_cell_elements
        return ValuesCell(
            present=True, values=vals[:cap], total=len(vals), truncated=len(vals) > cap
        )
    if p["kind"] == "element":
        eid = p["id"]
        return ElementCell(element_id=eid if eid in model.elements else None)
    ids = [i for i in dict.fromkeys(p["ids"]) if i in model.elements]
    cap = limits.max_cell_elements
    return ElementsCell(element_ids=ids[:cap], total=len(ids), truncated=len(ids) > cap)


def evaluate_cells(
    mm: Metamodel,
    model: Model,
    defn: TableDefinition,
    keys: list[RowKey],
    limits: TableLimits = TableLimits(),
    script: ScriptEvalContext | None = None,
) -> list[list[Cell]]:
    """Evaluate every cell for every row. Runs per page (unlike `build_rows`,
    which must see the whole table because expansion determines row count).

    `script` is the shared per-request `ScriptEvalContext` — `None` for
    callers with no script columns in play (every existing caller)."""
    expand_count = sum(
        1 for c in defn.columns if getattr(c, "mode", "collapse") == "expand"
    )
    # Correct here (unlike mid-`build_rows`) because `keys` is the FULL,
    # already-built set: base_slots = total slots minus one per expand column.
    base_slots = (len(keys[0]) - expand_count) if keys else 1

    rows: list[list[Cell]] = []
    for key in keys:
        row: list[Cell] = []
        for i, col in enumerate(defn.columns):
            if isinstance(col, ElementColumn):
                row.append(
                    _element_cell(mm, model, defn, key, col, base_slots, limits, script)
                )
            elif isinstance(col, PropertyColumn):
                row.append(
                    _property_cell(
                        mm, model, defn, key, col, i, base_slots, limits, script
                    )
                )
            elif isinstance(col, ScriptColumn):
                row.append(
                    _script_cell(
                        mm, model, defn, key, col, i, base_slots, limits, script
                    )
                )
            else:
                # NavigationColumn — the last Column-union member.
                assert isinstance(col, NavigationColumn)
                row.append(
                    _navigation_cell(
                        mm, model, defn, key, col, i, base_slots, limits, script
                    )
                )
        rows.append(row)
    return rows

File: engine/src/table/cells.ts
/**
 * A table's cells, as `evaluate_cells` of `core/table/cells.py` evaluates them
 * and `POST /tables/evaluate` writes them: every field present, in the
 * route's order, an element as its tree row. Values stay exact here; the page
 * puts them on the wire. A collapse cell reads its column's source afresh; an
 * expand cell reads back the slot the build promoted into its row's key.
 */
import type { Model } from '../model/model.ts';
import { displayName } from '../model/naming.ts';
import { PropertyValue, type Meter } from '../navigation/evaluate.ts';
import { treeItem, type TreeItem } from '../read/tree.ts';
import type { Steps } from '../steps/steps.ts';
import type { Value } from '../value/types.ts';
import type { NavMemo } from './nav-memo.ts';
import {
	elementIds,
	expandSlotOf,
	navigationReached,
	resolveSourceElements,
	type Binding,
	type Pass,
	type RowKey,
	type TableLimits
} from './rows.ts';
import type {
	ElementColumn,
	NavigationColumn,
	PropertyColumn,
	ScriptColumn,
	TableDefinition
} from './schema.ts';
import {
	isVirtualProperty,
	propertyDatatype,
	propertyDeclared,
	propertyIsElementTyped,
	rawProperty
} from './virtual-props.ts';

/** One cell, as the route's `TableCellOut` carries it. */
export type TableCell = {
	kind: 'element' | 'value' | 'values' | 'elements' | 'error';
	/** element: the element referred to. */
	item: TreeItem | null;
	/** element: the type an editable reference's picker offers. */
	ref_type: string | null;
	present: boolean | null;
	value: Value;
	/** value: the element holding it; element: the owner of the reference. */
	element_id: string | null;
	editable: boolean | null;
	items: TreeItem[] | null;
	values: Value[] | null;
	total: number | null;
	truncated: boolean | null;
	message: string | null;
	traceback: string | null;
};

function elementCell(
	model: Model,
	id: string | null,
	owner: string | null = null,
	editable = false,
	refType: string | null = null
): TableCell {
	return {
		kind: 'element',
		item: id === null ? null : treeItem(model, model.getElement(id)),
		ref_type: refType,
		present: null,
		value: null,
		element_id: owner,
		editable,
		items: null,
		values: null,
		total: null,
		truncated: null,
		message: null,
		traceback: null
	};
}

function valueCell(
	present: boolean,
	value: Value,
	elementId: string | null,
	editable: boolean
): TableCell {
	return {
		kind: 'value',
		item: null,
		ref_type: null,
		present,
		value,
		element_id: elementId,
		editable,
		items: null,
		values: null,
		total: null,
		truncated: null,
		message: null,
		traceback: null
	};
}

const emptyValueCell = (): TableCell => valueCell(false, null, null, false);

function valuesCell(values: Value[], total: number, truncated: boolean): TableCell {
	return {
		kind: 'values',
		item: null,
		ref_type: null,
		present: true,
		value: null,
		element_id: null,
		editable: null,
		items: null,
		values,
		total,
		truncated,
		message: null,
		traceback: null
	};
}

function elementsCell(model: Model, ids: string[], total: number, truncated: boolean): TableCell {
	return {
		kind: 'elements',
		item: null,
		ref_type: null,
		present: null,
		value: null,
		element_id: null,
		editable: null,
		items: ids.map((id) => treeItem(model, model.getElement(id))),
		values: null,
		total,
		truncated,
		message: null,
		traceback: null
	};
}

/** A slot's value; a value terminal's is the value it carries. */
const slotValue = (b: Binding): Value => (b instanceof PropertyValue ? b.value : b);

/** An id naming an element, else nothing. */
const elementRef = (model: Model, value: Value): string | null =>
	typeof value === 'string' && model.findElement(value) !== undefined ? value : null;

/**
 * What a collapse property column holds over `owners`: lists flattened,
 * `null` skipped, an owner whose type does not declare it contributing nothing.
 */
function propertyValues(pass: Pass, col: PropertyColumn, owners: readonly string[]): Value[] {
	const values: Value[] = [];
	for (const id of owners) {
		const element = pass.model.getElement(id);
		if (!propertyDeclared(pass.mm, element.typeName, col.name)) continue;
		const v = rawProperty(element, col.name);
		if (Array.isArray(v)) values.push(...v);
		else if (v !== undefined && v !== null) values.push(v);
	}
	return values;
}

function* elementColumnCell(pass: Pass, key: RowKey, col: ElementColumn): Steps<TableCell> {
	const ids = yield* resolveSourceElements(pass, key, col.source);
	return elementCell(pass.model, ids[0] ?? null);
}

/**
 * An element-typed property on any source element renders its ids as
 * elements, dropping the scalars other types hold under the same name. One
 * source element's value is editable where its type declares the property.
 */
function* propertyCell(
	pass: Pass,
	key: RowKey,
	col: PropertyColumn,
	index: number
): Steps<TableCell> {
	const { model, mm, defn, baseSlots } = pass;
	const owners = yield* resolveSourceElements(pass, key, col.source);
	const elementTyped = owners.some((id) =>
		propertyIsElementTyped(mm, model.getElement(id).typeName, col.name)
	);
	if (col.mode === 'expand') {
		const value = slotValue(key[expandSlotOf(defn, baseSlots, index)]!);
		// A value of many source elements has no one owner.
		const owner = owners.length === 1 ? owners[0]! : null;
		if (elementTyped) return elementCell(model, elementRef(model, value), owner, false);
		const present = owners.some((id) =>
			propertyDeclared(mm, model.getElement(id).typeName, col.name)
		);
		return valueCell(present, value, owner, false);
	}
	if (owners.length === 0) return emptyValueCell();
	if (owners.length === 1) {
		const id = owners[0]!;
		const element = model.getElement(id);
		const present = propertyDeclared(mm, element.typeName, col.name);
		const value = present ? (rawProperty(element, col.name) ?? null) : null;
		if (elementTyped && !Array.isArray(value)) {
			const refType = propertyDatatype(mm, element.typeName, col.name);
			return elementCell(model, elementRef(model, value), id, true, refType);
		}
		if (elementTyped) {
			const ids = elementIds(model, value);
			return elementsCell(model, ids, ids.length, false);
		}
		return valueCell(present, value, id, present && !isVirtualProperty(col.name));
	}
	const values = propertyValues(pass, col, owners);
	if (elementTyped) {
		const ids = elementIds(model, values);
		return elementsCell(model, ids, ids.length, false);
	}
	return valuesCell(values, values.length, false);
}

/**
 * A collapse navigation cell shows at most `min(cell_cap, maxCellElements)`
 * of what it reaches, or `maxCellElements` when the limits ignore cell caps,
 * `total` counting them all. A value anywhere makes it a values cell, an
 * element among values showing as its name.
 */
function* navigationCell(
	pass: Pass,
	key: RowKey,
	col: NavigationColumn,
	index: number,
	limits: TableLimits
): Steps<TableCell> {
	const { model, defn, baseSlots } = pass;
	if (col.mode === 'expand') {
		const b = key[expandSlotOf(defn, baseSlots, index)];
		if (b instanceof PropertyValue) return valueCell(true, b.value, null, false);
		return elementCell(model, typeof b === 'string' ? b : null);
	}
	const roots = yield* resolveSourceElements(pass, key, col.source);
	const [reached] = yield* navigationReached(pass, col, roots);
	const cap =
		limits.ignoreCellCaps === true
			? limits.maxCellElements
			: Math.min(col.cell_cap, limits.maxCellElements);
	if (reached.some((node) => node instanceof PropertyValue)) {
		const values = reached.map((node) =>
			typeof node === 'string' ? displayName(model.getElement(node)) : node.value
		);
		return valuesCell(values.slice(0, cap), values.length, values.length > cap);
	}
	const ids = reached as string[];
	return elementsCell(model, ids.slice(0, cap), ids.length, ids.length > cap);
}

/** A script column here runs no snippet: a configured one reaches a script and is refused before any cell. */
function scriptCell(pass: Pass, key: RowKey, col: ScriptColumn, index: number): TableCell {
	if (col.snippet.ref !== null || col.snippet.definition !== null) {
		throw new Error('a script column with a snippet is evaluated on the server');
	}
	if (col.mode === 'expand') {
		const b = key[expandSlotOf(pass.defn, pass.baseSlots, index)];
		if (b instanceof PropertyValue) return valueCell(true, b.value, null, false);
		if (typeof b === 'string') return elementCell(pass.model, b);
	}
	return emptyValueCell();
}

/**
 * The cells of `keys`, a row of cells per key, over a definition resolved and
 * reaching no script. `baseSlots` is the build's; `memo` is this pass's own.
 */
export function* evaluateCellsSteps(
	model: Model,
	defn: TableDefinition,
	keys: readonly RowKey[],
	baseSlots: number,
	limits: TableLimits,
	meter: Meter,
	memo: NavMemo | null
): Steps<TableCell[][]> {
	const pass: Pass = { model, mm: model.metamodel, defn, baseSlots, meter, memo };
	const rows: TableCell[][] = [];
	for (const key of keys) {
		const row: TableCell[] = [];
		for (const [index, col] of defn.columns.entries()) {
			switch (col.kind) {
				case 'element':
					row.push(yield* elementColumnCell(pass, key, col));
					break;
				case 'property':
					row.push(yield* propertyCell(pass, key, col, index));
					break;
				case 'navigation':
					row.push(yield* navigationCell(pass, key, col, index, limits));
					break;
				case 'script':
					row.push(scriptCell(pass, key, col, index));
					break;
			}
			if (meter.tick()) yield meter.end();
		}
		rows.push(row);
	}
	return rows;
}
